toml/src/main/kotlin/org/rust/toml/crates/local/CratesLocalIndexService.kt

```diff
@@ -0,0 +1,274 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.toml.crates.local
+
+import com.google.gson.Gson
+import com.intellij.openapi.Disposable
+import com.intellij.openapi.components.PersistentStateComponent
+import com.intellij.openapi.components.State
+import com.intellij.openapi.components.Storage
+import com.intellij.openapi.diagnostic.Logger
+import com.intellij.openapi.diagnostic.logger
+import com.intellij.openapi.progress.ProgressIndicator
+import com.intellij.openapi.progress.Task
+import com.intellij.util.io.*
+import org.eclipse.jgit.lib.ObjectId
+import org.eclipse.jgit.lib.Repository
+import org.eclipse.jgit.revwalk.RevTree
+import org.eclipse.jgit.revwalk.RevWalk
+import org.eclipse.jgit.storage.file.FileRepositoryBuilder
+import org.eclipse.jgit.treewalk.TreeWalk
+import org.rust.openapiext.pluginDirInSystem
+import org.rust.stdext.cleanDirectory
+import org.rust.toml.crates.local.CratesLocalIndexService.Companion.CratesLocalIndexState
+import java.io.*
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.Paths
+import java.util.concurrent.atomic.AtomicBoolean
+
+/**
+ * Crates local index, created from user cargo registry index on host machine.
+ * Used for dependency code insight in project's `Cargo.toml`.
+ * Stores crates info in [crates] persistent hash map and hash for commit which has been used for index load in
+ * persistent state [CratesLocalIndexState].
+ */
+@State(name = "CratesLocalIndexState", storages = [Storage("rust.crateslocalindex.xml")])
+class CratesLocalIndexService : PersistentStateComponent<CratesLocalIndexState>, Disposable {
+    private val userCargoIndexDir: Path
+        get() = Paths.get(System.getProperty("user.home"), CARGO_REGISTRY_INDEX_LOCATION)
+
+    // TODO: handle RepositoryNotFoundException
+    private val repository: Repository = FileRepositoryBuilder()
+        .setGitDir(userCargoIndexDir.toFile())
+        .build()
+
+    private val registryHeadCommitHash: String = repository.resolve("origin/master")?.name ?: run {
+        LOG.error("Git revision string cannot be resolved to any object id")
+        INVALID_COMMIT_HASH
+    }
+
+    private val crates: PersistentHashMap<String, CargoRegistryCrate>? = run {
+        resetIndexIfNeeded()
+
+        val file = baseCratesLocalRegistryDir.resolve("crates-local-index")
+        try {
+            IOUtil.openCleanOrResetBroken({
+                PersistentHashMap(
+                    file,
+                    EnumeratorStringDescriptor.INSTANCE,
+                    CrateExternalizer,
+                    4 * 1024,
+                    CRATES_INDEX_VERSION
+                )
+            }, file)
+        } catch (e: IOException) {
+            LOG.error("Cannot open or create PersistentHashMap in $file", e)
+            null
+        }
+    }
+
+    @Volatile
+    private var state: CratesLocalIndexState = CratesLocalIndexState()
```

Comment:
Why do we have a mutable variable with mutable state? It makes `@Volatile` useless